### ISSUE TITLE
WSREP: FSM: no such a transition JOINED -> JOINED

### ISSUE THREAD

**Issue by evanskinner** [codership/galera#420]
Running these software versions on Debian 7:

```
ii  libperconaserverclient18.1         5.6.30-25.16-1.wheezy          amd64        Percona XtraDB Cluster database client library
ii  percona-xtrabackup                 2.3.5-1.wheezy                 amd64        Open source backup tool for InnoDB and XtraDB
ii  percona-xtradb-cluster-client-5.6  5.6.30-25.16-1.wheezy          amd64        Percona XtraDB Cluster database client binaries
ii  percona-xtradb-cluster-common-5.6  5.6.30-25.16-1.wheezy          amd64        Percona XtraDB Cluster database common files (e.g. /etc/mysql/my.cnf)
ii  percona-xtradb-cluster-galera-3    3.16-1.wheezy                  amd64        Metapackage for latest version of galera3.
ii  percona-xtradb-cluster-galera-3.x  3.16-1.wheezy                  amd64        Galera components of Percona XtraDB Cluster
ii  percona-xtradb-cluster-server-5.6  5.6.30-25.16-1.wheezy          amd64        Percona XtraDB Cluster database server binaries
ii  xtrabackup                         2.3.5-1.wheezy                 all          Transitional package for percona-xtrabackup

```

I bootstrapped my 3 node cluster and all was going well, but after the final node got an SST the Donor node for that SST crashed with signal 6 and the following output in syslog:

```
Aug 15 17:11:11 apimdev0189 -innobackupex-backup: 160815 17:11:11 [01] Streaming ./mediumorg2016_11/file_managed.frm to <STDOUT>
Aug 15 17:11:11 apimdev0189 -innobackupex-backup: 160815 17:11:11 [01]        ...done
Aug 15 17:11:11 apimdev0189 -innobackupex-backup: 160815 17:11:11 [01] Streaming ./mediumorg2016_11/field_revision_application_description.frm to <STDOUT>
Aug 15 17:11:11 apimdev0189 -innobackupex-backup: 160815 17:11:11 [01]        ...done
Aug 15 17:11:11 apimdev0189 -innobackupex-backup: 160815 17:11:11 [01] Streaming ./mediumorg2016_11/entity_translation.frm to <STDOUT>
Aug 15 17:11:11 apimdev0189 rsyslogd-2177: imuxsock begins to drop messages from pid 30953 due to rate-limiting
Aug 15 17:11:22 apimdev0189 rsyslogd-2177: imuxsock lost 505 messages from pid 30953 due to rate-limiting
Aug 15 17:11:22 apimdev0189 -innobackupex-backup: 160815 17:11:22 Executing UNLOCK TABLES
Aug 15 17:11:22 apimdev0189 mysqld: 2016-08-15 17:11:22 6676 [Note] WSREP: resuming provider at 28480
Aug 15 17:11:22 apimdev0189 mysqld: 2016-08-15 17:11:22 6676 [Note] WSREP: Provider resumed.
Aug 15 17:11:22 apimdev0189 -innobackupex-backup: 160815 17:11:22 All tables unlocked
Aug 15 17:11:22 apimdev0189 -innobackupex-backup: 160815 17:11:22 Backup created in directory '/tmp/tmp.qpZTT3aFv0'
Aug 15 17:11:22 apimdev0189 -innobackupex-backup: MySQL binlog position: filename 'mysql-bin.000005', position '56653949'
Aug 15 17:11:22 apimdev0189 -innobackupex-backup: 160815 17:11:22 [00] Streaming backup-my.cnf
Aug 15 17:11:22 apimdev0189 -innobackupex-backup: 160815 17:11:22 [00]        ...done
Aug 15 17:11:22 apimdev0189 mysqld: 2016-08-15 17:11:22 6676 [Note] WSREP: 1.0 (apimdev0189.hursley.ibm.com): State transfer to 0.0 (apimdev0190.hursley.ibm.com) complete.
Aug 15 17:11:22 apimdev0189 mysqld: 2016-08-15 17:11:22 6676 [Note] WSREP: Shifting DONOR/DESYNCED -> JOINED (TO: 59544924)
Aug 15 17:11:22 apimdev0189 -innobackupex-backup: 160815 17:11:22 [00] Streaming xtrabackup_info
Aug 15 17:11:22 apimdev0189 -innobackupex-backup: 160815 17:11:22 [00]        ...done
Aug 15 17:11:22 apimdev0189 -innobackupex-backup: xtrabackup: Transaction log of lsn (119284128026) to (119382706803) was copied.
Aug 15 17:11:22 apimdev0189 -innobackupex-backup: 160815 17:11:22 completed OK!
Aug 15 17:11:22 apimdev0189 -wsrep-sst-donor: Total time on donor: 0 seconds
Aug 15 17:11:22 apimdev0189 -wsrep-sst-donor: Cleaning up temporary directories
Aug 15 17:11:22 apimdev0189 mysqld: 2016-08-15 17:11:22 6676 [Note] WSREP: 1.0 (apimdev0189.hursley.ibm.com): State transfer to 0.0 (apimdev0190.hursley.ibm.com) complete.
Aug 15 17:11:35 apimdev0189 mysqld: 2016-08-15 17:11:35 6676 [Note] WSREP: 0.0 (apimdev0190.hursley.ibm.com): State transfer from 1.0 (apimdev0189.hursley.ibm.com) complete.
Aug 15 17:11:35 apimdev0189 mysqld: 2016-08-15 17:11:35 6676 [ERROR] WSREP: FSM: no such a transition JOINED -> JOINED
Aug 15 17:11:35 apimdev0189 mysqld: 17:11:35 UTC - mysqld got signal 6 ;
Aug 15 17:11:35 apimdev0189 mysqld: This could be because you hit a bug. It is also possible that this binary
Aug 15 17:11:35 apimdev0189 mysqld: or one of the libraries it was linked against is corrupt, improperly built,
Aug 15 17:11:35 apimdev0189 mysqld: or misconfigured. This error can also be caused by malfunctioning hardware.
Aug 15 17:11:35 apimdev0189 mysqld: We will try our best to scrape up some info that will hopefully help
Aug 15 17:11:35 apimdev0189 mysqld: diagnose the problem, but since we have already crashed, 
Aug 15 17:11:35 apimdev0189 mysqld: something is definitely wrong and this may fail.
Aug 15 17:11:35 apimdev0189 mysqld: Please help us make Percona XtraDB Cluster better by reporting any
Aug 15 17:11:35 apimdev0189 mysqld: bugs at https://bugs.launchpad.net/percona-xtradb-cluster
Aug 15 17:11:35 apimdev0189 mysqld: 
Aug 15 17:11:35 apimdev0189 mysqld: key_buffer_size=134217728
Aug 15 17:11:35 apimdev0189 mysqld: read_buffer_size=2097152
Aug 15 17:11:35 apimdev0189 mysqld: max_used_connections=7
Aug 15 17:11:35 apimdev0189 mysqld: max_threads=502
Aug 15 17:11:35 apimdev0189 mysqld: thread_count=20
Aug 15 17:11:35 apimdev0189 mysqld: connection_count=3
Aug 15 17:11:35 apimdev0189 mysqld: It is possible that mysqld could use up to 
Aug 15 17:11:35 apimdev0189 mysqld: key_buffer_size + (read_buffer_size + sort_buffer_size)*max_threads = 3222581 K  bytes of memory
Aug 15 17:11:35 apimdev0189 mysqld: Hope that's ok; if not, decrease some variables in the equation.
Aug 15 17:11:35 apimdev0189 mysqld: 
Aug 15 17:11:35 apimdev0189 mysqld: Thread pointer: 0x2468040
Aug 15 17:11:35 apimdev0189 mysqld: Attempting backtrace. You can use the following information to find out
Aug 15 17:11:35 apimdev0189 mysqld: where mysqld died. If you see no messages after this, something went
Aug 15 17:11:35 apimdev0189 mysqld: terribly wrong...
Aug 15 17:11:35 apimdev0189 mysqld: stack_bottom = 7f071817ba90 thread_stack 0x40000
Aug 15 17:11:35 apimdev0189 mysqld: /usr/sbin/mysqld(my_print_stacktrace+0x2e)[0x907d1e]
Aug 15 17:11:35 apimdev0189 mysqld: /usr/sbin/mysqld(handle_fatal_signal+0x491)[0x6c4fa1]
Aug 15 17:11:35 apimdev0189 mysqld: /lib/x86_64-linux-gnu/libpthread.so.0(+0xf0a0)[0x7f08407f80a0]
Aug 15 17:11:35 apimdev0189 mysqld: /lib/x86_64-linux-gnu/libc.so.6(gsignal+0x35)[0x7f083e825125]
Aug 15 17:11:35 apimdev0189 mysqld: /lib/x86_64-linux-gnu/libc.so.6(abort+0x180)[0x7f083e8283a0]
Aug 15 17:11:35 apimdev0189 mysqld: /usr/lib/libgalera_smm.so(_ZN6galera3FSMINS_10Replicator5StateENS_13ReplicatorSMM10TransitionENS_10EmptyGuardENS3_11StateActionEE8shift_toES2_+0x1ee
)[0x7f07dd03b49e]
Aug 15 17:11:35 apimdev0189 mysqld: /usr/lib/libgalera_smm.so(_ZN6galera13ReplicatorSMM12process_joinEll+0x7c)[0x7f07dd03135c]
Aug 15 17:11:35 apimdev0189 mysqld: /usr/lib/libgalera_smm.so(_ZN6galera15GcsActionSource8dispatchEPvRK10gcs_actionRb+0x6cd)[0x7f07dd00e93d]
Aug 15 17:11:35 apimdev0189 mysqld: /usr/lib/libgalera_smm.so(_ZN6galera15GcsActionSource7processEPvRb+0x5c)[0x7f07dd00fabc]
Aug 15 17:11:35 apimdev0189 mysqld: /usr/lib/libgalera_smm.so(_ZN6galera13ReplicatorSMM10async_recvEPv+0x7f)[0x7f07dd031f8f]
Aug 15 17:11:35 apimdev0189 mysqld: /usr/lib/libgalera_smm.so(galera_recv+0x24)[0x7f07dd041744]
Aug 15 17:11:35 apimdev0189 mysqld: /usr/sbin/mysqld[0x61b785]
Aug 15 17:11:35 apimdev0189 mysqld: /usr/sbin/mysqld(start_wsrep_THD+0x28a)[0x601e1a]
Aug 15 17:11:35 apimdev0189 mysqld: /lib/x86_64-linux-gnu/libpthread.so.0(+0x6b50)[0x7f08407efb50]
Aug 15 17:11:35 apimdev0189 mysqld: /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d)[0x7f083e8d0fbd]
Aug 15 17:11:35 apimdev0189 mysqld: 
Aug 15 17:11:35 apimdev0189 mysqld: Trying to get some variables.
Aug 15 17:11:35 apimdev0189 mysqld: Some pointers may be invalid and cause the dump to abort.
Aug 15 17:11:35 apimdev0189 mysqld: Query (0): is an invalid pointer
Aug 15 17:11:35 apimdev0189 mysqld: Connection ID (thread ID): 5
Aug 15 17:11:35 apimdev0189 mysqld: Status: NOT_KILLED
Aug 15 17:11:35 apimdev0189 mysqld: 
Aug 15 17:11:35 apimdev0189 mysqld: You may download the Percona XtraDB Cluster operations manual by visiting
Aug 15 17:11:35 apimdev0189 mysqld: http://www.percona.com/software/percona-xtradb-cluster/. You may find information
Aug 15 17:11:35 apimdev0189 mysqld: in the manual which will help you identify the cause of the crash.
Aug 15 17:11:35 apimdev0189 mysqld_safe: Number of processes running now: 0
Aug 15 17:11:35 apimdev0189 mysqld_safe: WSREP: not restarting wsrep node automatically
Aug 15 17:11:35 apimdev0189 mysqld_safe: mysqld from pid file /var/run/mysqld/mysqld.pid ended
```

I am guessing this bug is in Galera but of course I maybe be wrong and it might be in xtrabackup or XtraDB. Please just let me know if it is and I will move the bug over to them. Thanks.

**Comment by evanskinner** [comment]
Note that this is an intermittent error. Sometimes the entire cluster bootstraps fine.

**Comment by philip-galera** [comment]
Can you provide the complete logs from the donor and the joiner?

**Comment by evanskinner** [comment]
Sure. The complete syslog?

More info: In the case above. Once I restarted the crashed node, it got an SST from another node and then that Donor node crashed with the same bug. Once I then restarted that node it got an IST and now all the nodes are Primary.

**Comment by evanskinner** [comment]
Here are the syslogs for both machines.

[apimdev0189.syslog.gz](https://github.com/codership/galera/files/419755/apimdev0189.syslog.gz)
[apimdev0190.syslog.gz](https://github.com/codership/galera/files/419756/apimdev0190.syslog.gz)

I since shutdown and re-bootstrapped and everything got an SST just fine this time without any crashes.

I did a major version upgrade and you'll see in the logs when XtraDB is starting and complaining that some tables have incorrect columns. After that I ran mysql_upgrade to fix it. I don;t see why that would be related but flagging it up anyway.
